gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt

```diff
@@ -88,233 +142,123 @@ open class AnvilPlugin : KotlinCompilerPluginSupportPlugin {
+            // needed for merging Dagger modules.
+            //
+            // However, we still need to update the generated code when the compiler plugin classpath
+            // changes.
+            disableIncrementalCompilationWhenTheCompilerClasspathChanges(
+              variant, incrementalSignal, stubsTaskProvider
             )
+          } else {
+            stubsTaskProvider.configure { stubsTask ->
+              // Disable incremental compilation for the stub generating task. Trigger the compiler
+              // plugin if any dependencies in the compile classpath have changed. This will make sure
+              // that we pick up any change from a dependency when merging all the classes. Without
+              // this workaround we could make changes in any library, but these changes wouldn't be
+              // contributed to the Dagger graph, because incremental compilation tricked us.
+              stubsTask.doFirstCompat {
+                stubsTask.incremental = false
+                stubsTask.logger.info(
+                  "Anvil: Incremental compilation enabled: ${stubsTask.incremental} (stub)"
+                )
+              }
+            }
           }
         }
     }
-
-    if (projectType.isAndroid) {
-      project.androidVariantsConfigure { variant ->
-        val compileTaskName = "compile${variant.name.capitalize(US)}$taskSuffix"
-        disableIncrementalCompilationAction(project, incrementalSignal, compileTaskName)
-      }
-    } else {
-      // The Java plugin has two Kotlin tasks we care about: compileKotlin and compileTestKotlin.
-      disableIncrementalCompilationAction(project, incrementalSignal, "compile$taskSuffix")
-      disableIncrementalCompilationAction(project, incrementalSignal, "compileTest$taskSuffix")
-    }
   }
 
-  private fun disableIncrementalCompilationAction(
-    project: Project,
+  private fun disableIncrementalCompilationWhenTheCompilerClasspathChanges(
+    variant: Variant,
     incrementalSignal: Provider<IncrementalSignal>,
-    compileTaskName: String
+    compileTaskProvider: TaskProvider<out KotlinCompile>
   ) {
+    val compileTaskName = compileTaskProvider.name
+
     // Disable incremental compilation, if the compiler plugin dependency isn't up-to-date.
     // This will trigger a full compilation of a module using Anvil even though its
     // source files might not have changed. This workaround is necessary, otherwise
     // incremental builds are broken. See https://youtrack.jetbrains.com/issue/KT-38570
-    val disableIncrementalCompilationTaskProvider = project.tasks.register(
+    val disableIncrementalCompilationTaskProvider = variant.project.tasks.register(
       compileTaskName + "CheckIncrementalCompilationAnvil",
       DisableIncrementalCompilationTask::class.java
     ) { task ->
       task.pluginClasspath.from(
-        project.configurations.getByName(PLUGIN_CLASSPATH_CONFIGURATION_NAME)
```

Comment:
Here was a very tricky bug. The code is correct with Kotlin 1.4, because there was only a single Kotlin compiler plugin classpath. What this code does is to disable incremental compilation, if the compiler plugin classpath changed. This is for example important when we bump the Anvil version, because new or different might need to be generated.

With Kotlin 1.5 this classpath is empty and they generate compiler plugin classpaths for every single variant (main, test, android test, debug, release, etc.) and add our compiler plugin to these specific classpaths. But here we kept watching the old empty classpath, but there can never be a change. 

That was tricky to track down :-/

Reply:
Wait for the next PR that became a lot simpler because of that.